### PR TITLE
[op-conductor] part 2 - core control logic

### PR DESCRIPTION
**Pull request by 0x00101010**
**Description**

Core control logic of op-conductor, it comprises of the functionality to handle sequencer state transition based on different states.

**Tests**

There are 2 sets of tests added
1. success scenarios (which contains 7 scenarios for normal status transition)
2. failure and retry scenarios

**Metadata**

- https://github.com/ethereum-optimism/protocol-quest/issues/47

### CODE REVIEW

File: op-conductor/conductor/service_test.go [hunk 1]
```diff
@@ -17,20 +20,20 @@ import (
...
 	return Config{
 		ConsensusAddr:  "127.0.0.1",
 		ConsensusPort:  50050,
 		RaftServerID:   "SequencerA",
-		RaftStorageDir: dir,
+		RaftStorageDir: "/tmp/raft",
 		RaftBootstrap:  false,
 		NodeRPC:        "http://node:8545",
 		ExecutionRPC:   "http://geth:8545",
+		Paused:         false,
```
[the remark this answers is not in the document]
Comment (0x00101010) [reply op-conductor/conductor/service_test.go:36]: This does not matter since consensus is mocked

File: op-conductor/conductor/service.go [hunk 2]
```diff
@@ -349,5 +358,119 @@ func (oc *OpConductor) action() {
...
+		// start sequencer
+		err = oc.startSequencer()
+	case status.leader && status.healthy && status.active:
+		// normal leader, do nothing
+	}
+
+	if err != nil {
+		oc.log.Error("failed to execute step, queueing another one to retry", "err", err)
+		// randomly sleep for 0-200ms to avoid excessive retry
+		time.Sleep(time.Duration(rand.Intn(200)) * time.Millisecond)
```
Comment (zhwrd) [inline op-conductor/conductor/service.go:406]: do you think it's worth making this sleep interval configurable?

**Review by zhwrd — APPROVED** [review]
otherwise looks good to me!

File: op-conductor/conductor/service.go (hunk 2 above)
Comment (0x00101010) [reply op-conductor/conductor/service.go:406]: I thought about this, the main thing is that we don't want to have retry time to be too long (longer than block time), but we don't want it to be too short (excessive retry), so currently this random 0-200ms feels good to me.

My thought is that we can debate on a time that we're all comfortable with, but I don't see a lot of benefit to having this configurable. Let me know if it makes sense

**Pull request merged**